GVFS/GVFS.Common/Tracing/TelemetryDaemonEventListener.cs

```diff
@@ -0,0 +1,163 @@
+            };
+
+            return JsonConvert.SerializeObject(message);
+        }
+
+        protected override void RecordMessageInternal(
+            string eventName,
+            Guid activityId,
+            Guid parentActivityId,
+            EventLevel level,
+            Keywords keywords,
+            EventOpcode opcode,
+            string payload)
+        {
+            string message = CreateJsonMessage(
+                this.vfsVersion,
+                this.providerName,
+                this.enlistmentId,
+                this.mountId,
+                eventName,
+                activityId,
+                parentActivityId,
+                level,
+                keywords,
+                opcode,
+                payload);
+
+            this.SendMessage(message);
+        }
+
+        private static string GetConfigValue(string gitBinRoot, string configKey)
+        {
+            string value = string.Empty;
+            string error;
+
+            GitProcess.ConfigResult result = GitProcess.GetFromSystemConfig(gitBinRoot, configKey);
+            if (!result.TryParseAsString(out value, out error, defaultValue: string.Empty) || string.IsNullOrWhiteSpace(value))
+            {
+                result = GitProcess.GetFromGlobalConfig(gitBinRoot, configKey);
+                result.TryParseAsString(out value, out error, defaultValue: string.Empty);
+            }
+
+            return value.TrimEnd('\r', '\n');
+        }
+
+        private void SendMessage(string message)
+        {
+            // Create pipe if this is the first message, or if the last connection broke for any reason
+            if (this.pipeClient == null)
+            {
+                string pipeName = GVFSPlatform.Instance.GetTelemetryNamedPipeName();
+                var pipe = new NamedPipeClientStream(".", pipeName, PipeDirection.Out, PipeOptions.Asynchronous);
+                try
+                {
+                    // Specify a instantaneous timeout because we don't want to hold up the rest of the
+                    // application if the pipe is not available; we will just drop this event.
+                    pipe.Connect(timeout: 0);
+                    this.pipeClient = pipe;
+                }
+                catch (TimeoutException)
+                {
+                    // We can't connect; we will try again with a new pipe on the next message
+                    return;
+                }
+            }
+
+            try
+            {
+                var buffer = Encoding.UTF8.GetBytes(message);
+                this.pipeClient.Write(buffer, 0, buffer.Length);
+            }
+            catch (Exception)
+            {
+                // We can't send this message for some reason (e.g., broken pipe); we attempt no recovery or retry
+                // mechanism and drop this message. We will try to recreate/connect the pipe on the next message.

```

Comment:
So if someone decides to turn off the daemon we will lose all messages?  And we will try the connect every time?  Do you know what the perf impact will be?  I know you have the timeout at `0` but there is still a cost at the very least at the exception being thrown.  Wondering if we should have a message queue with a limited size and try connecting and writing messages on a background thread?  I know that is added complexity but would get all this code out of the way in performance critical code paths.

Reply:
> So if someone decides to turn off the daemon we will lose all messages?

Yes. If the mount process (or service.. or whoever created the `JsonTracer`) was started when the `gvfs.telemetry-id` configuration entry was set then telemetry tracing is on for the lifetime of the process. We don't re-evaluate the setting on failure (say if they disabled it after mounting).

> And we will try the connect every time?

Yes, for every message (when telemetry is enabled at start). We _could_ recheck the telemetry configuration entry on failure to send a message, and then stop sending them until it was set again? But again checking the config every telemetry event would equally add some overhead (possibly even more as it would involve a git.exe process start).

> Do you know what the perf impact will be? I know you have the timeout at 0 but there is still a cost at the very least at the exception being thrown.

I did some very rough profiling and I could only measure 0-1 ms delay per message. The sending of a successful message was also between 0-1 ms per message however. I can and will dig deeper on these numbers however.

> Wondering if we should have a message queue with a limited size and try connecting and writing messages on a background thread? I know that is added complexity but would get all this code out of the way in performance critical code paths.

An interesting idea, but yes this would add more complexity. Do we think it's worth it? Do we want to risk breaking due to failing to send telemetry, or risk losing events if case of a problem?

Reply:
I'd like for us to trace once that we were not able to connect to the daemon if we expected to be able to, for supportability.

Reply:
See John's comment about logging an error here.

Reply:
> I'd like for us to trace once that we were not able to connect to the daemon if we expected to be able to, for supportability.

Hmm.. that'd be interesting to do, seeing as we're in the event listener tracing code itself.
I'd need to, as an `EventListener` have a reference to the current `ITracer`, and trace to it (in the hope that the log-file listener has been added.. then I'd need to ensure I don't get stuck in a recursive loop attempting and failing to trace the message to telemetry that the telemetry tracing didn't work 🤓

Reply:
If there isn't a noticeable performance impact then this should be fine but like John said it would be nice if we had some record that we are dropping messages.

Reply:
@jrbriggs @kewillford @jeffhostetler, I've pushed another commit which now traces the first instance of an `EventListener` failure to all other listeners (and any new listeners that are registered after the failure).

Please have another look at that most recent commit.